### PR TITLE
added oneauth project

### PR DESCRIPTION
**Pull request by aljaxus**

### CODE REVIEW

**Comment by greenstatic** [comment]
> Izpostavljena je bila kriticna varnostna luknja v pristopu

What was the critical security hole? :)

**Comment by aljaxus** [comment]
> What was the critical security hole? :)
> *- @greenstatic*

The issue is QR code proxy-ing on phishing pages. Now instead of getting the user's username and password all you needed was for them to scan the QR code and you had full account access.


https://twitter.com/PirateSoftware/status/1216527198863015937

https://portswigger.net/daily-swig/discord-users-warned-over-qr-code-login-scam-that-can-result-in-pwned-accounts

---

The attacker would create a phishing page attacking users of ServiceA. Client1 would fall for some email scam and open the PhishingA page. PhishingA page would contact ServiceA, start a new session (Session1) and request a QR code in order to authenticate. PhishingA would then show that QR code to Client1 - who thinks that he's looking at ServiceA. Client1 would scan the QR code with Client1phone. Client1phone would then contact ServiceA, authenticate and authorize Session1 to Client1account.

The problem is that not only Client1 had access to Session1, but also the PhishingA system had access to it (well to the [jwt](https://www.npmjs.com/package/jsonwebtoken) that is used to authenticate to ServiceA as Session1).

Meaning that PhishingA now has access to Session1, which is authenticated as Client1account. And Client1 is now compromised.

---

The project had basically the same approach as Discord has, but for a single app and multiple server-side libraries, project plugins (eg. wordpress plugin). Meaning the the security issue is basically the same.

discord only providers this authentication approach with their app, for their platform.  This project would provide one app (for clients) and libraries (for devs) used on different services (personal websites, etc...)

Discord still hasn't fixed this issue, as it's not really possible to fix it without requiring any user input - which completely defeats the goal of the project - which was to remove the user input from the authentication flow, except the QRcode scan from the user's phone.

I guess the excitement and pigeon-hole-view "prevented" me from seeing this issue earlier.

---

It was brought to my attention by @craftbyte

**Comment by craftbyte** [comment]
Yep, my fault 😅 . We realized only FIDO WebAuthn is the correct solution, especially with the new PassKeys.

**Comment by greenstatic** [comment]
Thanks for the detailed answer :)

So what you are essentially describing is a classical man-in-the-middle attack. We already know a couple of techniques against this.

I’m sure you already thought of mitigations and came up empty handed, so I’m intellectually curious to see where is the technical issue.

Let me bounce off some ideas which you have probably thought off and learn where the issue is - so we can all learn from this experience :)

I’m not familiar with your protocol but I assume you are doing some sort of challenge-response type of thing. The QR code probably contains some challenge and the client upon scanning the QR code sends the response back to the same URL. This of course being prone to man in the middle attacks. 

1. How far off am I?

I see two issues:

Issue A: You are sending the secret back in the challenge response (albeit on a pseudo-“secure” channel (i.e. TLS) but you have no idea with who you are communicating (e.g. attacker)).

Issue B: You are trusting the URL of the QR code.

You probably have something quite elementary like a static secret key (i.e. randomly generated password?). What about some form of zero-knowledge proof? This is just sorta good hygiene in the protocol, solving issue A.

Issue B can be solved a step before. During pairing/registration of a device (phone) you are by definition on a trusted site. So you could exchange a URL with which to ALWAYS use for authentication. This would render the man in the middle attack useless because you are communicating the proof (not secret if using zero-knowledge proof) over the trusted authentication URL from the pairing/registration process and NOT the untrusted data from the QR code.
You even have a feature now, since the trusted site knows which challenges it created you could detect man in the middle attacks, thus warn the user they are being attacked and to close the site with the QR code.

These two ideas I just came up on the top of my head reading you response. 

2. What am I missing, where do these two things go wrong and bight you in the rear end? Of course ignoring the whole “don’t invent your own crypto thing”.


3. Oh, and how are you combatting replay attacks?

**Comment by aljaxus** [comment]
> The QR code probably contains some challenge and the client upon scanning the QR code sends the response back to the same URL

No. The URL API endpoint URL is stored in the mobile app storage from the initial "registration" of the account (registratin in this context is when the client "links" their mobile app the their online account on a given platform. Just like TOTP apps do). The login QR code does contain the API endpoint url, but the client does not use any arbitrary URL, it checks if the API url is already present in the local store (meaning the API URL is actually a trusted one).


> Issue A: You are sending the secret back in the challenge response (albeit on a pseudo-“secure” channel (i.e. TLS) but you have no idea with who you are communicating (e.g. attacker)).

The mobile app knows exactly to who it's talking with when communicating with the service API.

The problem is that the mobile app does not have any way of knowing if the QR code is being served on the service page (correct) or if it's being "proxied" on a phishing page (attacked).

> Issue B: You are trusting the URL of the QR code.

Half-trusting. It is used in order to "identify" any already added "profiles" in the mobile app. It is not implicitly trusted.


> Issue B can be solved a step before. During pairing/registration of a device (phone) you are by definition on a trusted site. So you could exchange a URL with which to ALWAYS use for authentication. This would render the man in the middle attack useless because you are communicating the proof (not secret if using zero-knowledge proof) over the trusted authentication URL from the pairing/registration process and NOT the untrusted data from the QR code.
You even have a feature now, since the trusted site knows which challenges it created you could detect man in the middle attacks, thus warn the user they are being attacked and to close the site with the QR code.

All of this was thought of. The problem is that an attacker can just replicate the client browser UI on their phishing page and communicate with the Service API as if they are the real client, and communicate with the real client as if they are the Service that the Client want to authenticate to.

> 3. Oh, and how are you combatting replay attacks?

TOTP would be one of the challenges in the authentication flow


In order to solve this issue, there would have to be a way for the mobile app to verify that the webpage that the client has open in their browser is actually, without a doubt, the webpage of the service that the user wants to authenticate to.

If there was a way to verify that the Service API endpoint is indeed the one that the QR code is presenting to be, then the issue would be resolved, but there is no way for the mobile app to communicate with the client browser and verify that the API endpoints / domains are all valid.

At least that I know of. I guess we would implement the feature where the user scans the QR code and then also scans the URL bar in the browser, but some browsers (chrome) show only domains, and not full URLs. This is also not secure enough, because phishing pages usually have very similar domains to original services' domains, and text recognition may recognize the phishing domain as valid.

We also can not rely on page-generated content (eg. a javascript scriptlet integrating the page domain into the QR code and then display that one) because the attacker can just generate arbitrary content and again pretend that they are the `service-a.com`, while they are `soevlce-a.com` (one is lowercase `I` other is lowecase `L`).

---

TL;DR 

In order to make this system work securely, there would have to be a way for the browser itself (which we know is a trusted entity) somehow reliably tell the mobile app which website is currently served.

**Comment by greenstatic** [comment]
Glad you thought about all this. Good stuff.

I see, you don't have a backend library but an actual HTTP API. In that case what prevents something like an OAuth 2 Authorization Code Grant flow?

**Comment by aljaxus** [comment]
> I see, you don't have a backend library but an actual HTTP A

Well no.. I'm talking about APIs because that's what the mobile app would be talking with.

The project would develop and maintain libraries for different webservers in different languages which would implement this standardized API interface.

Developers would then use these libraries to simply and *- khm -* securely implement authentication to their applications without worrying about usernames and passwords, or just use the system as a second factor.

---

I guess we could "prevent" (not really, more like just restrict cheap attackers) by requiring a valid reCaptcha token passed with the request when the browser requests the QR code.
This would somehow limit the attackers from simply proxy-ing QR codes.

Though there are enough automated systems that solve all sorts of captchas nowdays...
And this would also defeat the main goal - erasing the need of direct user interaction with the browser in the login flow.

**Comment by greenstatic** [comment]
I might be missing something or my point is not getting across. If we have something that resembles the OAuth 2 Auth Code Flow but modified a bit with QR codes instead of passwords. In what scenario does a man in the middle attack break this?

(I haven't thought this through entirely, but you I'm sure have so I'm curious 😄)

Screen Shot 2022-10-13 at 5 31 31 PM

**Comment by aljaxus** [comment]
I am not exactly totally familiar with oAuth2 flow specifics

> In what scenario does a man in the middle attack break this?

As said before;

I could create a phishing site, when you come and "try to login" I start a new session with the actual leggit service, I get the QR code and session details, I present you the QR code and interact with the API as if I was a legitimate client, you talk with the phishing site thinking that they are the legitimate server.

There is no way to verify that the browser is on the correct page and not on a phishing page.

This is also what FIDO apparently solves.

**Comment by greenstatic** [comment]
Sorry, not convinced yet 😅.

The new Slovenian eletronic government issued identity cards will have the feature of (in slovene), [from the FAQ](https://www.si-trust.gov.si/sl/podpora-uporabnikom/pogosta-vprasanja/elektronska-osebna-izkaznica/) under "Kako bom uporabljal mobilno aplikacijo?":
> uporaba na drugi napravi npr. tablici ali računalniku. Uporabnik prične postopek e-storitve na spletni strani ponudnika e-storitve na svojem računalniku ali tablici in pri izboru ustreznega načina prijave izbere prijavo z mobilno aplikacjo in e-osebno izkaznico. Na zaslonu se izpiše koda QR. Nato uporabnik odpre na svojem pametnem telefonu mobilno aplikacijo s katero skenira kodo QR in na mobilni aplikacji sledi postopku vnosa kode PIN in branja elektronske osebne izkaznice preko protokola NFC. Po uspešni avtentikaciji je uporabnik prijavljen in nadaljuje opravljanje e-storitve na npr. tablici ali računalniku.

Besides the reading of the ID card using the smart phone, how is this (apparently very similar flow) immune to the attack you supposedly describe?

**Comment by craftbyte** [comment]
It is not, I already tested with some test gear :P
The QR code is a JWT which tells the app the backend channel to use. I dont
even know why they use a JWT.

**Comment by greenstatic** [comment]
@craftbyte Will you create a PoC when the app will be officially released? 😄

**Comment by craftbyte** [comment]
Heh, if I find the time.

**Comment by aljaxus** [comment]
> Besides the reading of the ID card using the smart phone, how is this (apparently very similar flow) immune to the attack you supposedly describe?

Sadly enough there's many systems that are vulnerable to such ... basic? ... attacks and are still deployed to production and kept there. A prime example is Discord.
*I guess the devs/project leads don't have the balls to admit the system is flawed and that they should remove it from prod.*
